### PR TITLE
 Adding AuthenticationFailedException to Azure.Identity fixes issue #6712

### PR DESCRIPTION
**Pull request by schaabs**

### CODE REVIEW

File: sdk/identity/Azure.Identity/src/AadIdentityClient.cs [hunk 1]
```diff
@@ -28,6 +28,8 @@ internal class AadIdentityClient
 
         private const string ClientAssertionType = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
 
+        private const string AuthenticationRequestFailedError = "The reqeust to the identity service failed.  See inner exception for details.";
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/AadIdentityClient.cs]: request

File: sdk/identity/Azure.Identity/src/AadIdentityClient.cs [hunk 2]
```diff
@@ -42,68 +44,87 @@ public virtual async Task<AccessToken> AuthenticateAsync(string tenantId, string
...
+
+            if (response.Status == 200 || response.Status == 201)
+            {
+                var result = await DeserializeAsync(response.ContentStream, cancellationToken).ConfigureAwait(false);
 
-                throw response.CreateRequestFailedException();
+                return new Response<AccessToken>(response, result);
             }
+
+            throw await response.CreateRequestFailedExceptionAsync();
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/AadIdentityClient.cs]: Does this require a .ConfigureAwait(false) (and everywhere we call something like that?)

File: sdk/identity/Azure.Identity/src/AuthenticationFailedException.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/AuthenticationFailedException.cs]: Are all of these required?

We need to add the copyright header here.

File: sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs [hunk 4]
```diff
@@ -30,9 +30,24 @@ public DefaultAzureCredential()
...
         public DefaultAzureCredential(IdentityClientOptions options)
-            : base(new EnvironmentCredential(options), new ManagedIdentityCredential(options: options))
+            : base(new EnvironmentCredential(options), new ManagedIdentityCredential(options: options), new CredentialNotFoundGuard())
         {
 
         }
+
+        private class CredentialNotFoundGuard : TokenCredential
+        {
+            private const string CredentialNotFoundMessage = @"Failed to find a credential to use for authentication.  If running in an environment where a managed identity is not available ensure AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET are set.";
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs]: I would say "ensure AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET **environment variables**

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 5]
```diff
@@ -15,7 +15,11 @@ namespace Azure.Identity
     internal class ManagedIdentityClient
     {
         private static Lazy<ManagedIdentityClient> s_sharedClient = new Lazy<ManagedIdentityClient>(() => new ManagedIdentityClient());
-        
+
+        private const string AuthenticationResponseInvalidFormatError = "Invalid response.  The authentication response was not in the expected format";
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Extra space after `Invalid response.`

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 6]
```diff
@@ -15,7 +15,11 @@ namespace Azure.Identity
...
+        private const string MsiEndpointInvalidUriError = "The environment variable MSI_ENDPOINT contains an invalid Uri";
+        private const string AuthenticationRequestFailedError = "The reqeust to the identity service failed.  See inner exception for details.";
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Extra space?

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 5 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Get off my lawn kid ;).

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 7]
```diff
@@ -16,9 +16,9 @@ internal class ManagedIdentityClient
     {
         private static Lazy<ManagedIdentityClient> s_sharedClient = new Lazy<ManagedIdentityClient>(() => new ManagedIdentityClient());
 
-        private const string AuthenticationResponseInvalidFormatError = "Invalid response.  The authentication response was not in the expected format";
+        private const string AuthenticationResponseInvalidFormatError = "Invalid response. The authentication response was not in the expected format";
         private const string MsiEndpointInvalidUriError = "The environment variable MSI_ENDPOINT contains an invalid Uri";
-        private const string AuthenticationRequestFailedError = "The reqeust to the identity service failed.  See inner exception for details.";
+        private const string AuthenticationRequestFailedError = "The reqeust to the identity service failed. See inner exception for details.";
```
Comment (AlexGhiondea) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: request

**Review by AlexGhiondea — APPROVED** [review]

**Pull request merged**